sled-agent/config-reconciler/src/dataset_serialization_task.rs

```diff
@@ -399,9 +456,140 @@ impl DatasetTask {
+            // errors! We won't be able to parse any dataset names that can't be
+            // represented by `DatasetKind`, which will always include at least
+            // the two parent datasets we asked for above (`zpool` and
+            // `zpool/crypt`).
+            //
+            // Our goal here is to prune datasets that _we_ created; we only
+            // create datasets that have representable `DatasetName`s, so this
+            // also acts as a safeguard that we should never delete datasets we
+            // don't fully understand.
+            let Ok(dataset) = DatasetName::from_str(&properties.name) else {
+                continue;
+            };
+
+            // Does this dataset have an ID set? If we created it, we expect it
+            // does, and we can easily check whether it still exists in our
+            // config. A dataset with a known `DatasetKind` without an ID set is
+            // pretty unexpected: we'll search through our config datasets to
+            // find a matching name; if we do, we'll assume it's that one.
+            let present_in_config = match properties.id {
+                Some(id) => datasets.contains_key(&id),
+                None => {
+                    if datasets.iter().any(|d| d.name == dataset) {
+                        warn!(
+                            self.log,
+                            "found on-disk dataset without an ID \
+                             that matches a config dataset by name";
+                            "dataset" => dataset.full_name(),
+                        );
+                        true
+                    } else {
+                        warn!(
+                            self.log,
+                            "found on-disk dataset without an ID \
+                             that doesn't match any config datasets; assuming \
+                             it should be marked as an orphan";
+                            "dataset" => dataset.full_name(),
+                        );
+                        false
+                    }
+                }
+            };
+            if present_in_config {
+                continue;
+            }
+
+            // Explicitly match all variants here, so if we expand this list
+            // we're forced to consider whether they should be removeable.
+            let reason = match dataset.kind() {
+                // Zone durable datasets: these are the kinds we expect to
+                // remove (e.g., when the corresponding zone is expunged).
+                DatasetKind::Cockroach
+                | DatasetKind::Crucible
+                | DatasetKind::Clickhouse
+                | DatasetKind::ClickhouseKeeper
+                | DatasetKind::ClickhouseServer
+                | DatasetKind::ExternalDns
+                | DatasetKind::InternalDns => {
+                    // We should attempt to delete this; for now, just report it
```

Comment:
I'm a little confused about the comments here - you're implying we "should delete" or "should refuse to delete", but this match statement just returns a "reason" with no label about whether deletion should proceed.

Is this a thing we're changing in a follow-up? Or a relic of a previous version?

Reply:
Changing in a followup: the `reason` for the dataset being orphaned in this branch is "we don't delete datasets yet (see omicron#6177)". Addressing #6177 therefore turns into replacing this with "try to delete the dataset, and only report an orphan reason if deletion fails".

The other dataset kinds will not be deleted as a part of the followup work to address #6177.

Reply:
Will open PR 3 of 2 momentarily that addresses this.